### PR TITLE
Preallocate segment files

### LINKED ISSUE #6504: InternalError: a fault occurred in a recent unsafe memory access operation in compiled Java code
**Describe the bug**
Observed in logs: https://console.cloud.google.com/errors/CMHr6Oj_nM3WDQ?service=zeebe&time=P7D&refresh=off&project=camunda-cloud-240911

**Log/Stacktrace**
<details><summary>Full Stacktrace</summary>
 <p>

```
java.lang.InternalError: a fault occurred in a recent unsafe memory access operation in compiled Java code
	at com.esotericsoftware.kryo.pool.SoftReferenceQueue.offer(SoftReferenceQueue.java:53) ~[kryo-4.0.2.jar:?]
	at com.esotericsoftware.kryo.pool.SoftReferenceQueue.offer(SoftReferenceQueue.java:33) ~[kryo-4.0.2.jar:?]
	at com.esotericsoftware.kryo.pool.KryoPoolQueueImpl.release(KryoPoolQueueImpl.java:52) ~[kryo-4.0.2.jar:?]
	at io.atomix.utils.serializer.NamespaceImpl.release(NamespaceImpl.java:321) ~[atomix-utils-0.26.1.jar:0.26.1]
	at io.atomix.utils.serializer.NamespaceImpl.serialize(NamespaceImpl.java:154) ~[atomix-utils-0.26.1.jar:0.26.1]
	at io.atomix.utils.serializer.FallbackNamespace.serialize(FallbackNamespace.java:71) ~[atomix-utils-0.26.1.jar:0.26.1]
	at io.atomix.storage.journal.MappedJournalSegmentWriter.append(MappedJournalSegmentWriter.java:112) ~[atomix-storage-0.26.1.jar:0.26.1]
	at io.atomix.storage.journal.SegmentedJournalWriter.append(SegmentedJournalWriter.java:54) ~[atomix-storage-0.26.1.jar:0.26.1]
	at io.atomix.storage.journal.DelegatingJournalWriter.append(DelegatingJournalWriter.java:46) ~[atomix-storage-0.26.1.jar:0.26.1]
	at io.atomix.raft.roles.PassiveRole.appendEntry(PassiveRole.java:710) ~[atomix-cluster-0.26.1.jar:0.26.1]
	at io.atomix.raft.roles.PassiveRole.appendEntry(PassiveRole.java:665) ~[atomix-cluster-0.26.1.jar:0.26.1]
	at io.atomix.raft.roles.PassiveRole.tryToAppend(PassiveRole.java:643) ~[atomix-cluster-0.26.1.jar:0.26.1]
	at io.atomix.raft.roles.PassiveRole.appendEntries(PassiveRole.java:587) ~[atomix-cluster-0.26.1.jar:0.26.1]
	at io.atomix.raft.roles.PassiveRole.handleAppend(PassiveRole.java:440) ~[atomix-cluster-0.26.1.jar:0.26.1]
	at io.atomix.raft.roles.ActiveRole.onAppend(ActiveRole.java:51) ~[atomix-cluster-0.26.1.jar:0.26.1]
	at io.atomix.raft.roles.FollowerRole.onAppend(FollowerRole.java:195) ~[atomix-cluster-0.26.1.jar:0.26.1]
	at io.atomix.raft.impl.RaftContext.lambda$registerHandlers$14(RaftContext.java:220) ~[atomix-cluster-0.26.1.jar:0.26.1]
	at io.atomix.raft.impl.RaftContext.lambda$runOnContext$21(RaftContext.java:231) ~[atomix-cluster-0.26.1.jar:0.26.1]
	at io.atomix.utils.concurrent.SingleThreadContext$WrappedRunnable.run(SingleThreadContext.java:188) [atomix-utils-0.26.1.jar:0.26.1]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
```

</p>
</details>

**Environment:**
 - OS: Linux
 - Zeebe Version: 0.26.1
 - Configuration: Camunda Cloud

### LINKED ISSUE #8099: SIGBUS error with 1.2.2
**Describe the bug**

Customer has reported that they got an SIGBUS error with the latest patch release and running their benchmark, it looks like it is happening in the journal.

Impact: Since the customer doesn't run on kubernetes the system had to be restarted. 

> https://www.geeksforgeeks.org/segmentation-fault-sigsegv-vs-bus-error-sigbus/
> 2) Bus Error (also known as SIGBUS and is usually signal 10) occur when a process is trying to access memory that the CPU cannot physically address.In other words the memory tried to access by the program is not a valid memory address.It caused due to alignment issues with the CPU (eg. trying to read a long from an address which isn’t a multiple of 4). SIGBUS is abbrivation for “Bus Error”.
> 
> SIGBUS signal occurs in below cases,
> -> Program instructs the CPU to read or write a specific physical memory address which is not valid / Requested physical address is unrecognized by the whole computer system.
> -> Unaligned access of memory (For example, if multi-byte accesses must be 16 bit-aligned, addresses (given in bytes) at 0, 2, 4, 6, and so on would be considered aligned and therefore accessible, while addresses 1, 3, 5, and so on would be considered unaligned.)

**To Reproduce**

I think we can ask them what they exactly doing, what I know is that they have 24 partitions and 4 brokers with replication factor 4.

**Expected behavior**

No sigbus I assume?

**Log/Stacktrace**

```
[36mzeebe_broker2 |[0m 2021-10-27 14:14:33.928 [] [raft-server-2-raft-partition-partition-6] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 33400058 (first log index: 33322496)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:33.971 [] [raft-server-2-raft-partition-partition-22] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 33226084 (first log index: 33091065)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:33.977 [] [raft-server-2-raft-partition-partition-24] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 51869258 (first log index: 51728804)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:33.984 [] [raft-server-2-raft-partition-partition-10] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 34657647 (first log index: 34533361)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:33.999 [] [raft-server-2-raft-partition-partition-6] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 33400058 (first log index: 33322496)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:34.045 [] [raft-server-2-raft-partition-partition-22] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 33226084 (first log index: 33091065)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:34.048 [] [raft-server-2-raft-partition-partition-24] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 51869258 (first log index: 51728804)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:34.061 [] [raft-server-2-raft-partition-partition-10] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 34657647 (first log index: 34533361)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:34.070 [] [raft-server-2-raft-partition-partition-6] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 33400058 (first log index: 33322496)
[36mzeebe_broker2 |[0m [thread 82 also had an error]
[36mzeebe_broker2 |[0m [thread 104 also had an error][thread 91 also had an error]
[36mzeebe_broker2 |[0m 
[36mzeebe_broker2 |[0m [thread 99 also had an error]
[36mzeebe_broker2 |[0m #
[36mzeebe_broker2 |[0m # A fatal error has been detected by the Java Runtime Environment:
[36mzeebe_broker2 |[0m #
[36mzeebe_broker2 |[0m #  [thread 83 also had an error]
[36mzeebe_broker2 |[0m [thread 93 also had an error]
[36mzeebe_broker2 |[0m [thread 49 also had an error][thread 48 also had an error]
[36mzeebe_broker2 |[0m 
[36mzeebe_broker2 |[0m SIGBUS (0x7) at pc=0x00007f91eb54d169, pid=8, tid=92
[36mzeebe_broker2 |[0m #
[36mzeebe_broker2 |[0m # JRE version: OpenJDK Runtime Environment 18.9 (11.0.13+8) (build 11.0.13+8)
[36mzeebe_broker2 |[0m # Java VM: OpenJDK 64-Bit Server VM 18.9 (11.0.13+8, mixed mode, sharing, tiered, compressed oops, g1 gc, linux-amd64)
[36mzeebe_broker2 |[0m # Problematic frame:
[36mzeebe_broker2 |[0m # v  ~StubRoutines::jbyte_disjoint_arraycopy
[36mzeebe_broker2 |[0m #
[36mzeebe_broker2 |[0m # Core dump will be written. Default location: /var/swiftcore/core.%e.8.%h.%t
[36mzeebe_broker2 |[0m #
[36mzeebe_broker2 |[0m # An error report file with more information is saved as:
[36mzeebe_broker2 |[0m # /usr/local/zeebe/hs_err_pid8.log
[36mzeebe_broker2 |[0m Compiled method (c2) 103585123 19667   !   4       io.camunda.zeebe.journal.file.SegmentedJournalWriter::append (54 bytes)
[36mzeebe_broker2 |[0m  total in heap  [0x00007f91f4170a90,0x00007f91f4174fe8] = 17752
[36mzeebe_broker2 |[0m  relocation     [0x00007f91f4170c08,0x00007f91f4170da0] = 408
[36mzeebe_broker2 |[0m  main code      [0x00007f91f4170da0,0x00007f91f41730a0] = 8960
[36mzeebe_broker2 |[0m  stub code      [0x00007f91f41730a0,0x00007f91f4173120] = 128
[36mzeebe_broker2 |[0m  oops           [0x00007f91f4173120,0x00007f91f4173148] = 40
[36mzeebe_broker2 |[0m  metadata       [0x00007f91f4173148,0x00007f91f4173340] = 504
[36mzeebe_broker2 |[0m  scopes data    [0x00007f91f4173340,0x00007f91f4174738] = 5112
[36mzeebe_broker2 |[0m  scopes pcs     [0x00007f91f4174738,0x00007f91f4174dd8] = 1696
[36mzeebe_broker2 |[0m  dependencies   [0x00007f91f4174dd8,0x00007f91f4174e00] = 40
[36mzeebe_broker2 |[0m  handler table  [0x00007f91f4174e00,0x00007f91f4174ef0] = 240
[36mzeebe_broker2 |[0m  nul chk table  [0x00007f91f4174ef0,0x00007f91f4174fe8] = 248
[36mzeebe_broker2 |[0m Could not load hsdis-amd64.so; library not loadable; PrintAssembly is disabled
[36mzeebe_broker2 |[0m #
[36mzeebe_broker2 |[0m # If you would like to submit a bug report, please visit:
[36mzeebe_broker2 |[0m #   https://bugreport.java.com/bugreport/crash.jsp
[36mzeebe_broker2 |[0m #
[36mzeebe_broker2 exited with code 134
```

[hs_err_pid8.log](https://github.com/camunda-cloud/zeebe/files/7442605/hs_err_pid8.log)

**Environment:**
 - OS: 
 - Zeebe Version: 1.2.2
 - Configuration: 4 broker, 24 partitions, bare metal

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR introduces segment file pre-allocation in the journal. This is on by default, but can be disabled via an experimental configuration option.

At the moment, the pre-allocation is done in a "dumb" fashion - we allocate a 4Kb blocks of zeroes, and write this until we've reached the expected file length. Note that this means there may be one extra block allocated on disk.

One thing to note, to verify this, we used [jnr-posix](https://github.com/jnr/jnr-posix). The reason behind this is we want to know the actual number of blocks on disk reserved for this file. `Files#size`, or `File#length`, return the reported file size, which is part of the file's metadata (on UNIX systems anyway). If you mmap a file with a size of 1Mb, write one byte, then flush it, the reported size will be 1Mb, but the actual size on disk will be a single block (on most modern UNIX systems anyway). By using [stat](https://linux.die.net/man/2/stat), we can get the actual file size in terms of 512-bytes allocated blocks, so we get a pretty accurate measurement of the actual disk space used by the file.

I would've like to capture this in a test utility, but since `test-util` depends on `util`, there wasn't an easy way to do this, so I just copied the method in two places. One possibility I thought of is moving the whole pre-allocation stuff in `journal`, since we only use it there. The only downside I can see there is about discovery and cohesion, but I'd like to hear your thoughts on this.

A follow-up PR will come which will optimize the pre-allocation by using the [posix_fallocate](https://man7.org/linux/man-pages/man3/posix_fallocate.3.html) on POSIX systems.

Finally, I opted for an experimental configuration option instead of a feature flag. My reasoning is that it isn't a "new" feature, but instead we want to option of disabling this (for performance reasons potentially). So it's more of an advanced option. But I'd also like to hear your thoughts here.

## Related issues

closes #6504
closes #8099
related to #7607  

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by npepinpe** [comment]
NOTE: we might want to wait for https://github.com/camunda/zeebe/pull/9714 to be merged before reviewing this, as I expect some major conflicts :smile:

**Comment by deepthidevaki** [comment]
The pr is merged. Sorry for the conflicts :smile:

**Comment by npepinpe** [comment]
Happy to split off the last commit to a separate PR :+1:

**Comment by npepinpe** [comment]
ping @oleschoenburg let me know if you're too busy to have a look at this

File: journal/src/main/java/io/camunda/zeebe/journal/file/SegmentLoader.java [hunk 1]
```diff
@@ -165,9 +165,18 @@ private byte readVersion(final FileChannel channel, final String fileName) throw
...
+            StandardOpenOption.READ,
+            StandardOpenOption.WRITE,
+            StandardOpenOption.CREATE_NEW)) {
+      if (preallocateFiles) {
+        FileUtil.preallocate(channel, descriptor.maxSegmentSize());
+      }
+
+      return mapSegment(channel, descriptor.maxSegmentSize());
     } catch (final FileAlreadyExistsException e) {
       // assuming we haven't written in that segment, just overwrite it; if we have, we may be able
```
Comment (oleschoenburg) [inline journal/src/main/java/io/camunda/zeebe/journal/file/SegmentLoader.java:190]: :x: This comment is now outdated, we always delete if the file already exists.

File: journal/src/main/java/io/camunda/zeebe/journal/file/SegmentLoader.java [hunk 2]
```diff
@@ -10,39 +10,46 @@
...
+      final JournalIndex journalIndex) {
     final MappedByteBuffer mappedSegment;
 
     try {
-      mappedSegment = mapNewSegment(segmentFile, descriptor);
+      mappedSegment = mapNewSegment(segmentFile, descriptor, lastWrittenIndex);
     } catch (final IOException e) {
-      throw new JournalException(String.format("Failed to map new segment %s", segmentFile), e);
+      throw new JournalException(
+          String.format("Failed to allocate new segment file %s", segmentFile), e);
```
Comment (oleschoenburg) [inline journal/src/main/java/io/camunda/zeebe/journal/file/SegmentLoader.java:52]: :wrench: I suppose it'd be evident from the stacktrace but the message could be adjusted depending on `preallocateFiles`

**Review by lenaschoenburg — APPROVED** [review]
Very nice, only minor adjustments, nothing that would require a re-review :+1: 

:wrench: I wouldn't mind moving `preallocate` to the journal module. It'd save the duplicated `getRealSize` test helper and I general I think we don't expect to use this for anything else other than the journal any time soon.

:thought_balloon: Our example config files don't document this flag (or any other flag in `ExperimentalRaftCfg`). Not sure if we really want to add them since they are experimental after all.

:wrench: There are no tests for the new config, we could add them to `ExperimentalCfgTest`.

File: util/src/test/java/io/camunda/zeebe/util/FileUtilTest.java [hunk 3]
```diff
@@ -90,4 +97,60 @@ void shouldCopySnapshot() throws Exception {
...
+      FileUtil.preallocate(channel, length);
+    }
+
+    // then
+    final var realSize = getRealSize(path);
+    assertThat(realSize)
+        .as(
+            "Expected <%s> to have a real size of <%d> bytes, but it had <%d>",
+            path, length, realSize)
+        .isEqualTo(length);
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply util/src/test/java/io/camunda/zeebe/util/FileUtilTest.java:119]: Yes, you're 100% right, I thought about it, but was just very lazy :see_no_evil: It works as is as long as the size is a multiple of your I/O block size :sweat_smile:

File: journal/src/main/java/io/camunda/zeebe/journal/file/SegmentLoader.java (hunk 2 above)
Comment (npepinpe) [reply journal/src/main/java/io/camunda/zeebe/journal/file/SegmentLoader.java:52]: Any suggestions? The only thing I can imagine is the word `allocate` might alarm users who've turned off pre-allocation. So we could reword this to, say, create. That's fairly neutral, and as you said, the cause will give you the correct error depending on whether you're preallocating or not.

Or did you mean that users would want it explicit here whether they are pre-allocating or not?

File: util/src/test/java/io/camunda/zeebe/util/FileUtilTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (npepinpe) [reply util/src/test/java/io/camunda/zeebe/util/FileUtilTest.java:119]: Ah, I remember, the main issue is that you'll be between the real size and at most one block most. If we assume block size is 4096, all good - if we don't though, then this will fail. So it's a little better, but I guess still not really accurate. 

I could push it down and instead of `getRealSize` (which is more flexible), we just have a `assertThatFileHasRealSize()` and `assertThatFileHasRealSizeLessThan()`. When checking for equality, we can reuse the `FileStat#blockSize` to get an accurate upper bound.

[the remark this answers is not in the document]
Comment (npepinpe) [reply util/src/test/java/io/camunda/zeebe/util/FileUtilTest.java:119]: Actually, I just added a `getBlockSize`, it was less code in the end :shrug:

**Comment by npepinpe** [comment]
Re-review only for the tests comment, and to make sure we're aligned on https://github.com/camunda/zeebe/pull/9731#discussion_r918061292 (not sure I understood what you meant).

File: journal/src/main/java/io/camunda/zeebe/journal/file/SegmentLoader.java (hunk 2 above)
Comment (oleschoenburg) [reply journal/src/main/java/io/camunda/zeebe/journal/file/SegmentLoader.java]: I think we are on the same page here :+1:

File: util/src/test/java/io/camunda/zeebe/util/FileUtilTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (oleschoenburg) [reply util/src/test/java/io/camunda/zeebe/util/FileUtilTest.java]: Nice! Sorry, I didn't realize that this would be that much work, I just thought a `.isGreaterThan(length)` would have been enough but your version is obviously much stricter and better :+1:

**Review by lenaschoenburg — APPROVED** [review]
Still looks good ;)

Just to be sure, you have seen my comments here https://github.com/camunda/zeebe/pull/9731#pullrequestreview-1034386593 but ignored them, right? That's totally fine with me, but just wanted to make sure you didn't miss them.

**Comment by npepinpe** [comment]
No, sorry, I hadn't :see_no_evil:

**Comment by npepinpe** [comment]
I thought about moving the `preallocate` to the journal, and I might do that in the next PR as we pull in more native stuff. I'm not super keen on splitting file system utilities though, but I guess it is what it is.
